### PR TITLE
feat: update StoryState with v8 changes

### PR DESCRIPTION
**Pull request by ephread**
Thanks again for reviewing, we're getting there!

I have added a couple of comment, to highlight certain changes.

### CODE REVIEW

File: engine/StoryState.js [hunk 1]
```diff
@@ -312,31 +299,45 @@ export class StoryState{
...
+			var c = st.charAt(i);
+
+			var isInlineWhitespace = (c == ' ') || (c == '\t');
+
+			if (isInlineWhitespace && currentWhitespaceStart == -1)
+				currentWhitespaceStart = i;
+
+			if (!isInlineWhitespace) {
+				if (c != '\n' && currentWhitespaceStart > 0) {
+					sb.Append(str.substr(currentWhitespaceStart, i - currentWhitespaceStart));
```
Comment (ephread) [inline engine/StoryState.js]: This certainly needs a more thorough check, `substr` should do exactly the same thing as `Substring`, but we can't be too careful.

File: engine/StoryState.js [hunk 2]
```diff
@@ -385,15 +386,20 @@ export class StoryState{
...
+				this.OutputStreamDirty();
 				return;
 			}
 		}
 
 		this.PushToOutputStreamIndividual(obj);
 		this.OutputStreamDirty();
 	}
+	PopFromOutputStream(count){
+		this.outputStream.splice(this.outputStream.length - count, count);
```
Comment (ephread) [inline engine/StoryState.js]: Same here with `splice`, it should do exactly the same as `RemoveRange` but a double check would not be bad.

File: engine/StoryState.js [hunk 3]
```diff
@@ -518,31 +549,16 @@ export class StoryState{
...
-//			var glue = obj as Glue;
-			var glue = obj;
-
-			if (cmd instanceof ControlCommand || (txt instanceof StringValue && txt.isNonWhitespace)) {
-				foundNonWhitespace = true;
-				if( rightGlueToStopAt == null )
-					break;
-			} else if (rightGlueToStopAt && glue instanceof Glue && glue == rightGlueToStopAt) {
-				rightGluePos = i;
+			var cmd = (obj instanceof ControlCommand) ? obj : null;
```
Comment (ephread) [inline engine/StoryState.js]: Instead of specifying the type in a comment above and performing the test in the condition, I perform the test inline and assign accordingly, which mimics what the C# code does with `as`.

Let me know if should revert the change!

File: engine/StoryState.js [hunk 4]
```diff
@@ -597,49 +589,71 @@ export class StoryState{
...
+		// Debug.Assert (callStack.currentElement.type == PushPopType.Function);
+		var functionStartPoint = this.callStack.currentElement.functionStartInOutputStream;
+
+		if (functionStartPoint == -1) {
+			functionStartPoint = 0;
+		}
+
+		for (var i = this._outputStream.length - 1; i >= functionStartPoint; i--) {
+			var obj = this._outputStream[i];
+			var txt = (obj instanceof StringValue) ? obj : null;
```
Comment (ephread) [inline engine/StoryState.js]: Same as above (Line 556).

**Comment by ephread** [comment]
Just a heads up about the PR, as I've completed the port of `Story.cs`'s changes, I'm currently re-running the tests and fixing stuff as they break. I've already fixed a number of typos in `StoryState.js` and I'll update the PR when I'll be finished. \o/

File: engine/StoryState.js [hunk 5]
```diff
@@ -729,30 +742,31 @@ export class StoryState{
...
 		copy._currentChoices.push.apply(copy._currentChoices, this._currentChoices);
 
 		if (this.hasError) {
 			copy.currentErrors = [];
 			copy.currentErrors.push.apply(copy.currentErrors, this.currentErrors);
 		}
 
+		if (this.hasWarning) {
+			copy.currentWarnings = [];
+			copy.currentWarnings.push.apply(copy.currentWarnings, this.currentWarnings);
```
Comment (y-lohse) [inline engine/StoryState.js]: Thanks for sticking to the existing code style, but boy does that make me doubt my own sanity. A simple `copy. currentWarnings = this.currentWarnings.slice()` would have the same effect as far as I can tell?

File: engine/StoryState.js (hunk 3 above)
Comment (y-lohse) [reply engine/StoryState.js]: No that's actually way better, good idea 👍

File: engine/StoryState.js [hunk 6]
```diff
@@ -312,31 +299,45 @@ export class StoryState{
...
-				return g;
-			} else if (c instanceof ControlCommand) // e.g. BeginString
-				break;
+
+	CleanOutputWhitespace (str){
+		var sb = new StringBuilder(str.Length);
+
+		var currentWhitespaceStart = -1;
+
+		for (var i = 0; i < str.Length; i++) {
```
[the remark this answers is not in the document]
Comment (y-lohse) [reply engine/StoryState.js]: Argh, one of my comments didn't make it. There's another `str.Length` a few lines above.

File: engine/StoryState.js (hunk 3 above)
Comment (y-lohse) [reply engine/StoryState.js]: Nope that's actually way better, good idea 👍

**Comment by y-lohse** [comment]
By the way, feel free to start the PR for `Story.js`, or add it in this one if it's easier. I have a long train ride coming up on saturday so I'll have plenty of time to review things :)

File: engine/StoryState.js [hunk 7]
```diff
@@ -187,35 +179,33 @@ export class StoryState{
...
-	}
-	set currentPath(value){
-		if (value != null)
-			this.currentContentObject = this.story.ContentAtPath(value);
 		else
-			this.currentContentObject = null;
+			return pointer.path.toString();
+	}
+	get currentPointer(){
+		return this.callStack.currentElement.currentPointer.copy();
```
[the remark this answers is not in the document]
Comment (ephread) [reply engine/StoryState.js]: I took an extremely cautious approach, adding `copy()` everywhere I could think of and, even then, I still managed to shoot myself in the foot. I spend the last few hours debugging an infinite loop, only to discover that `Pointer.Null` wasn't returning a brand new object every time. Stupid, stupid oversight 😂.

As soon as all the test are passing again, we'll probably want to make a set of rules on how to handle value semantics in projects (i. e. which function has the responsibility of making the copy.)

**Comment by ephread** [comment]
Thanks for taking a look!

Most of the typos have now been fixed (thanks to the tests), but I'll updatethe PR later.

Story is massive, you probably want a separate PR. :) I'll also make an another PR which will fix a number of issues found in the other files I've updated. I'll try to keep everything clean and avoid a somewhat messed up history.

**Comment by ephread** [comment]
> I have a long train ride coming up on saturday so I'll have plenty of time to review things :)

Some monstrous diffs are waiting for your train ride! (#171, #172)
